
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Reconstructing atlas-level single-cell 3D tissue maps from serial sections">
      
      
      
        <link rel="canonical" href="https://a12910.github.io/space-map/examples/usage/">
      
      
        <link rel="prev" href="../../analysis/spatial-transcriptomics-analysis/">
      
      
        <link rel="next" href="../examples/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.13">
    
    
      
        <title>Usage Guide - Space-map</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.342714a4.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="blue">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#architecture-and-core-components" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Space-map" class="md-header__button md-logo" aria-label="Space-map" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Space-map
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Usage Guide
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/a12910/space-map" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    a12910/Space-map
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../overview/overview.md" class="md-tabs__link">
          
  
  
  Getting Started

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../architecture/architecture-and-core-components/" class="md-tabs__link">
          
  
  
  Architecture

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../registration/registration-system-%28lddmm%29/" class="md-tabs__link">
          
  
  
  Registration

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../data/data-management/" class="md-tabs__link">
          
  
  
  Data Management

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../workflow/key-workflows/" class="md-tabs__link">
          
  
  
  Workflows

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../analysis/spatial-transcriptomics-analysis/" class="md-tabs__link">
          
  
  
  Analysis

        </a>
      </li>
    
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="./" class="md-tabs__link">
          
  
  
  Examples

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../api/api/" class="md-tabs__link">
          
  
  
  API Reference

        </a>
      </li>
    
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../overview/contributing/" class="md-tabs__link">
        
  
  
    
  
  Contributing

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../overview/changelog/" class="md-tabs__link">
        
  
  
    
  
  Changelog

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Space-map" class="md-nav__button md-logo" aria-label="Space-map" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Space-map
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/a12910/space-map" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    a12910/Space-map
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Getting Started
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Getting Started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../overview/overview.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../overview/installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../overview/quickstart/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Quick Start
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Architecture
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Architecture
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../architecture/architecture-and-core-components/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Registration
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Registration
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../registration/registration-system-%28lddmm%29/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    LDDMM System
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../grid/grid-transformations/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Grid Transformations
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Data Management
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Data Management
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../data/data-management/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Flow Slice
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../feature/feature-matching/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Feature Matching
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Workflows
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Workflows
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../workflow/key-workflows/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Key Workflows
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../alignment/image-alignment/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Image Alignment
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../reconstruction/3d-reconstruction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3D Reconstruction
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Analysis
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            Analysis
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../analysis/spatial-transcriptomics-analysis/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Spatial Transcriptomics
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" checked>
        
          
          <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Examples
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_8">
            <span class="md-nav__icon md-icon"></span>
            Examples
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Usage Guide
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Usage Guide
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#system-architecture-overview" class="md-nav__link">
    <span class="md-ellipsis">
      System Architecture Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#core-components" class="md-nav__link">
    <span class="md-ellipsis">
      Core Components
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#registration-system" class="md-nav__link">
    <span class="md-ellipsis">
      Registration System
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grid-transformations" class="md-nav__link">
    <span class="md-ellipsis">
      Grid Transformations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-management-flow-slice" class="md-nav__link">
    <span class="md-ellipsis">
      Data Management (Flow, Slice)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#feature-matching" class="md-nav__link">
    <span class="md-ellipsis">
      Feature Matching
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#multi-resolution-strategy" class="md-nav__link">
    <span class="md-ellipsis">
      Multi-Resolution Strategy
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#transformdb-unified-transformation-management" class="md-nav__link">
    <span class="md-ellipsis">
      TransformDB: Unified Transformation Management
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#configuration-and-global-settings" class="md-nav__link">
    <span class="md-ellipsis">
      Configuration and Global Settings
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../examples/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Examples
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_9" >
        
          
          <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    API Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_9">
            <span class="md-nav__icon md-icon"></span>
            API Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/api/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../overview/contributing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Contributing
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../overview/changelog/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Changelog
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="architecture-and-core-components">Architecture and Core Components<a class="headerlink" href="#architecture-and-core-components" title="Permanent link">&para;</a></h1>
<p>This section provides a high-level overview of the SpaceMap architecture and its primary components. SpaceMap is designed as a modular framework for reconstructing 3D tissue maps from serial tissue sections, with a particular focus on spatial transcriptomics and proteomics data.</p>
<p>For detailed information about specific components, see the following sub-sections:
- <a href="../../registration/registration-system-%28lddmm%29/">Registration System (LDDMM)</a>
- <a href="../../grid/grid-transformations/">Grid Transformations</a>
- <a href="../../data/data-management/">Data Management (Flow, Slice)</a>
- <a href="../../feature/feature-matching/">Feature Matching</a></p>
<h2 id="system-architecture-overview">System Architecture Overview<a class="headerlink" href="#system-architecture-overview" title="Permanent link">&para;</a></h2>
<p>SpaceMap employs a layered architecture that separates data management, registration, and visualization concerns while maintaining efficient data flow between components.</p>
<h2 id="core-components">Core Components<a class="headerlink" href="#core-components" title="Permanent link">&para;</a></h2>
<p>SpaceMap consists of several interconnected components that work together to process and align tissue section data:
- <strong>FlowImport</strong>: Responsible for loading raw data (cell coordinates, images) and initializing the processing pipeline.
- <strong>Slice</strong>: Represents a single tissue section, containing both the raw data and any transformations applied to it.
- <strong>TransformDB</strong>: Stores and manages transformations between different sections.
- <strong>FlowExport</strong>: Handles exporting of aligned sections and 3D models.</p>
<h2 id="registration-system">Registration System<a class="headerlink" href="#registration-system" title="Permanent link">&para;</a></h2>
<p>The registration system is the core of SpaceMap, responsible for aligning tissue sections. It consists of two primary components:
1. <strong>Affine Registration</strong>: Provides coarse alignment by computing global transformations that account for translation, rotation, scaling, and shearing.
2. <strong>LDDMM Registration</strong>: Performs fine-grained, non-linear registration that preserves local structures while accurately aligning sections.</p>
<p>The two-stage approach combines computational efficiency with high accuracy, allowing SpaceMap to handle large datasets while maintaining precision.</p>
<p>For more detailed information about the registration process, see the <a href="../../registration/registration-system-%28lddmm%29/">Registration System</a> documentation.</p>
<h2 id="grid-transformations">Grid Transformations<a class="headerlink" href="#grid-transformations" title="Permanent link">&para;</a></h2>
<p>Grid transformations are a fundamental mechanism in SpaceMap for representing and applying non-linear deformations. These transformations are represented as regular grids where each point maps to another location, enabling complex warping operations.</p>
<p>Key features:
- Supports both forward and inverse transformations
- Provides bilinear interpolation for points between grid vertices
- Can be applied to both images and point clouds
- Facilitates combination of multiple transformations</p>
<h2 id="data-management-flow-slice">Data Management (Flow, Slice)<a class="headerlink" href="#data-management-flow-slice" title="Permanent link">&para;</a></h2>
<p>The data management components in SpaceMap handle the import, manipulation, and export of tissue section data. The primary classes are:
- <strong>FlowImport</strong>: Responsible for loading raw data (cell coordinates, images) and initializing the processing pipeline.
- <strong>Slice</strong>: Represents a single tissue section, containing both the raw data and any transformations applied to it.
- <strong>TransformDB</strong>: Stores and manages transformations between different sections.
- <strong>FlowExport</strong>: Handles exporting of aligned sections and 3D models.</p>
<p>For information about data management and preprocessing, refer to the <a href="../../data/data-management/">Data Management</a> guide.</p>
<h2 id="feature-matching">Feature Matching<a class="headerlink" href="#feature-matching" title="Permanent link">&para;</a></h2>
<p>Feature matching components are used to find correspondences between different tissue sections, which are then used as constraints in the registration process:
- <strong>SIFT Matching</strong>: Traditional computer vision approach for matching distinctive points between images.
- <strong>LOFTR (Learning Optical Flow from Transformers)</strong>: Deep learning-based method for feature matching, particularly effective for histological images.</p>
<h2 id="multi-resolution-strategy">Multi-Resolution Strategy<a class="headerlink" href="#multi-resolution-strategy" title="Permanent link">&para;</a></h2>
<p>SpaceMap employs a multi-resolution approach to registration to balance efficiency and accuracy. The <code>AutoFlowMultiCenter3</code> class in particular implements this strategy, enabling processing at multiple resolution levels for optimal results.</p>
<h2 id="transformdb-unified-transformation-management">TransformDB: Unified Transformation Management<a class="headerlink" href="#transformdb-unified-transformation-management" title="Permanent link">&para;</a></h2>
<p>The <code>TransformDB</code> class plays a central role in SpaceMap by providing a unified interface for storing and applying transformations:
- Supports both affine and grid-based transformations
- Forward and inverse transformations
- Automatic scaling to match input dimensions
- Batch processing for multiple images or point sets</p>
<h2 id="configuration-and-global-settings">Configuration and Global Settings<a class="headerlink" href="#configuration-and-global-settings" title="Permanent link">&para;</a></h2>
<p>SpaceMap uses global configuration settings to control various aspects of the system. These are defined in <code>base/root.py</code> and include:</p>
<table>
<thead>
<tr>
<th>Setting</th>
<th>Purpose</th>
<th>Default Value</th>
</tr>
</thead>
<tbody>
<tr>
<td>XYRANGE</td>
<td>Maximum coordinate range</td>
<td>4000</td>
</tr>
<tr>
<td>XYD</td>
<td>Scale factor for coordinate transforms</td>
<td>10</td>
</tr>
<tr>
<td>BASE</td>
<td>Base directory for data storage</td>
<td>"data/flow"</td>
</tr>
<tr>
<td>DEVICE</td>
<td>Computing device (CPU or GPU index)</td>
<td>"cpu" or 0</td>
</tr>
<tr>
<td>LAYER_START</td>
<td>Starting layer index</td>
<td>0</td>
</tr>
<tr>
<td>LAYER_END</td>
<td>Ending layer index</td>
<td>0</td>
</tr>
<tr>
<td>IMGCONF</td>
<td>Image configuration settings</td>
<td>{"raw": 1}</td>
</tr>
</tbody>
</table>
<p>The <code>auto_init_xyd</code> and <code>init_xy</code> functions in <code>base/base2.py</code> provide convenient ways to initialize these settings based on input data.</p>
<h1 id="key-workflows">Key Workflows<a class="headerlink" href="#key-workflows" title="Permanent link">&para;</a></h1>
<p>This section outlines the primary processing workflows in the SpaceMap framework for 3D tissue reconstruction. It provides a high-level overview of the main processing pipelines, their components, and how they interact. The workflows described here cover the entire process from raw data import to final 3D model generation and spatial analysis.</p>
<h2 id="main-workflows-overview">Main Workflows Overview<a class="headerlink" href="#main-workflows-overview" title="Permanent link">&para;</a></h2>
<p>SpaceMap implements a multi-stage approach to 3D tissue reconstruction, combining multi-scale feature matching with large-deformation diffeomorphic metric mapping (LDDMM). The core workflows are designed to balance computational efficiency with registration accuracy.</p>
<p><strong>Main Processing Pipeline:</strong></p>
<ol>
<li><strong>Data Import</strong> – Import and organize serial section data using <code>FlowImport</code>.</li>
<li><strong>Image Alignment</strong> – Multi-stage registration process to align neighboring sections using <code>AutoFlowMultiCenter2/3</code>.</li>
<li><strong>3D Reconstruction</strong> – Build 3D tissue models from aligned sections using <code>TransformDB</code>.</li>
<li><strong>Spatial Analysis</strong> – Analyze spatial transcriptomics data in the 3D context using <code>FlowExport</code>.</li>
</ol>
<h2 id="image-alignment-workflow">Image Alignment Workflow<a class="headerlink" href="#image-alignment-workflow" title="Permanent link">&para;</a></h2>
<p>The image alignment workflow is a core component of SpaceMap, combining coarse affine registration with fine LDDMM registration to accurately align tissue sections. This two-stage approach preserves both global structure and local micro-anatomy.</p>
<p><strong>Image Alignment Pipeline:</strong></p>
<ol>
<li><strong>Data Import:</strong> Raw data (cell coordinates and optional images) is imported using <code>FlowImport.init_xys()</code>, which organizes the data for processing.</li>
<li><strong>Slice Creation:</strong> The imported data is organized into <code>Slice</code> objects, which serve as the basic data structure for the alignment process.</li>
<li><strong>Coarse Alignment:</strong></li>
<li><strong>Affine Registration:</strong> Initial alignment using <code>AutoFlowMultiCenter2.affine()</code> to handle large displacements between sections.</li>
<li><strong>Feature Matching:</strong> Extraction of corresponding features between sections using SIFT or LOFTR for improved alignment.</li>
<li><strong>Fine Alignment:</strong></li>
<li><strong>LDDMM Registration:</strong> Local non-linear alignment using <code>AutoFlowMultiCenter2.ldm_pair()</code> or <code>AutoFlowMultiCenter2DF.ldm_pair()</code>.</li>
<li><strong>Grid Generation:</strong> Creation of transformation grids using <code>LDDMMRegistration.generate_img_grid()</code>.</li>
<li><strong>Grid Merging:</strong> Combination of transformation grids using <code>ldm_merge_pair()</code> for global consistency.</li>
</ol>
<p>The <code>AutoFlowMultiCenter2DF</code> class implements a multi-center approach that handles bidirectional alignment from a central slice, improving global consistency in the alignment.</p>
<h2 id="3d-reconstruction-workflow">3D Reconstruction Workflow<a class="headerlink" href="#3d-reconstruction-workflow" title="Permanent link">&para;</a></h2>
<p>The 3D reconstruction workflow transforms aligned 2D sections into a coherent 3D model by computing transformation grids and applying them to the original data.</p>
<p><strong>3D Reconstruction Pipeline:</strong></p>
<ol>
<li><strong>Grid Processing:</strong></li>
<li><strong>Grid Saving:</strong> Each alignment transformation is saved as a grid using <code>Slice.data.saveGrid()</code>.</li>
<li><strong>TransformDB Creation:</strong> A <code>TransformDB</code> object is created to store and manage all transformation grids.</li>
<li><strong>Grid Application:</strong> Transformation grids are applied to sections with <code>_apply_grid()</code>.</li>
<li><strong>3D Model Building:</strong></li>
<li><strong>Point Transformation:</strong> Cell coordinates are transformed using <code>TransformDB.apply_points()</code>.</li>
<li><strong>Model Export:</strong> The 3D model is exported with <code>FlowExport.export_imgs()</code> for visualization and analysis.</li>
</ol>
<p>The grid merging process is a key component of the 3D reconstruction workflow, implemented in <code>AutoFlowMultiCenter2DF.ldm_merge_pair()</code>. This method ensures that transformation grids are properly merged to maintain global consistency in the 3D reconstruction.</p>
<h2 id="spatial-transcriptomics-analysis-workflow">Spatial Transcriptomics Analysis Workflow<a class="headerlink" href="#spatial-transcriptomics-analysis-workflow" title="Permanent link">&para;</a></h2>
<p>The spatial transcriptomics analysis workflow enables analysis of gene expression patterns in the 3D reconstructed tissue.</p>
<p><strong>Spatial Analysis Pipeline:</strong></p>
<ol>
<li><strong>Data Analysis:</strong></li>
<li><strong>Cell Matrix Analysis:</strong> Processing of cell-gene expression matrices within the 3D context.</li>
<li><strong>Gene Expression Analysis:</strong> Analysis of gene expression patterns across the 3D tissue.</li>
<li><strong>Cell Type Identification:</strong> Classification of cells based on gene expression profiles.</li>
<li><strong>Visualization &amp; Export:</strong></li>
<li><strong>Spatial Statistics:</strong> Calculation of spatial relationships and patterns among cell types.</li>
<li><strong>3D Visualization:</strong> Interactive visualization of cells with gene expression data in 3D space.</li>
</ol>
<p>The <code>FlowExport</code> class handles the export of 3D models and transformed cell coordinates for downstream analysis.</p>
<h2 id="multi-center-workflow-strategy">Multi-center Workflow Strategy<a class="headerlink" href="#multi-center-workflow-strategy" title="Permanent link">&para;</a></h2>
<p>SpaceMap implements special multi-center workflows through the <code>AutoFlowMultiCenter2DF</code> and <code>AutoFlowMultiCenter3</code> classes. These workflows start from a central slice and work outward in both directions, which helps maintain global consistency in the reconstruction.</p>
<p><strong>Multi-center Alignment Strategy:</strong></p>
<ol>
<li><strong>Center Slice Selection:</strong> The middle slice is selected as a reference point.</li>
<li><strong>Bidirectional Processing:</strong> Slices are processed in both forward and backward directions from the center.</li>
<li><strong>Parallel Computation:</strong> Slice pairs can be processed in parallel for efficiency.</li>
<li><strong>Grid Merging:</strong> Transformation grids are merged to ensure global consistency.</li>
</ol>
<p>This bidirectional processing strategy helps maintain the global structure of the tissue while allowing for local deformations.</p>
<h2 id="configuration-options-and-parameters">Configuration Options and Parameters<a class="headerlink" href="#configuration-options-and-parameters" title="Permanent link">&para;</a></h2>
<p>The SpaceMap workflows can be customized through various configuration options and parameters:</p>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default</th>
<th>Component</th>
</tr>
</thead>
<tbody>
<tr>
<td>alignMethod</td>
<td>Method for initial alignment</td>
<td>"auto"</td>
<td>AutoFlowMultiCenter2</td>
</tr>
<tr>
<td>gpu</td>
<td>GPU device to use for computation</td>
<td>None</td>
<td>AutoFlowMultiCenter2</td>
</tr>
<tr>
<td>finalErr</td>
<td>Error threshold for LDDMM convergence</td>
<td>System default</td>
<td>ldm_pair</td>
</tr>
<tr>
<td>centerTrain</td>
<td>Whether to use center-based training</td>
<td>False</td>
<td>ldm_pair</td>
</tr>
<tr>
<td>show</td>
<td>Whether to display visual results</td>
<td>False</td>
<td>Various methods</td>
</tr>
<tr>
<td>saveGridKey</td>
<td>Key for saving transformation grids</td>
<td>pairGridKey</td>
<td>ldm_pair</td>
</tr>
</tbody>
</table>
<p>These parameters allow fine-tuning of the workflows for different types of data and reconstruction requirements.</p>
<h2 id="programming-against-the-workflow-system">Programming Against the Workflow System<a class="headerlink" href="#programming-against-the-workflow-system" title="Permanent link">&para;</a></h2>
<p>When developing with the SpaceMap framework, it's important to understand the main workflow components and how they work together. The following best practices are recommended:</p>
<ol>
<li><strong>Initialize with FlowImport:</strong> Always start by creating a <code>FlowImport</code> object to organize your data.</li>
<li><strong>Use the appropriate AutoFlow class:</strong> Select between <code>AutoFlowMultiCenter2</code>, <code>AutoFlowMultiCenter2DF</code>, or <code>AutoFlowMultiCenter3</code> depending on your alignment needs.</li>
<li><strong>Follow the workflow steps sequentially:</strong> First perform affine alignment, then LDDMM, then merge the grids.</li>
<li><strong>Use TransformDB for transformation management:</strong> Store all transformations in a <code>TransformDB</code> for efficient application.</li>
<li><strong>Export results with FlowExport:</strong> Use <code>FlowExport</code> to save your 3D models and transformed data.</li>
</ol>
<h1 id="detailed-usage-guide">Detailed Usage Guide<a class="headerlink" href="#detailed-usage-guide" title="Permanent link">&para;</a></h1>
<p>This guide provides detailed examples for using Space-map's core functionalities. Space-map employs a sophisticated two-stage registration approach that combines dimensionality reduction with robust feature matching to handle large-scale single-cell data efficiently.</p>
<h2 id="table-of-contents">Table of Contents<a class="headerlink" href="#table-of-contents" title="Permanent link">&para;</a></h2>
<ul>
<li><a href="#data-preparation">Data Preparation</a></li>
<li><a href="#registration-workflow">Registration Workflow</a></li>
<li><a href="#advanced-registration">Advanced Registration</a></li>
<li><a href="#working-with-large-datasets">Working with Large Datasets</a></li>
<li><a href="#visualizing-results">Visualizing Results</a></li>
<li><a href="#exporting-3d-models">Exporting 3D Models</a></li>
</ul>
<h2 id="data-preparation">Data Preparation<a class="headerlink" href="#data-preparation" title="Permanent link">&para;</a></h2>
<h3 id="supported-data-formats">Supported Data Formats<a class="headerlink" href="#supported-data-formats" title="Permanent link">&para;</a></h3>
<p>Space-map works with various data types:</p>
<div class="highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">spacemap</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sm</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">tifffile</span>

<span class="c1"># Cell coordinates (numpy array)</span>
<span class="n">points</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="n">x1</span><span class="p">,</span> <span class="n">y1</span><span class="p">],</span> <span class="p">[</span><span class="n">x2</span><span class="p">,</span> <span class="n">y2</span><span class="p">],</span> <span class="o">...</span><span class="p">])</span>  <span class="c1"># Shape: (n_cells, 2)</span>

<span class="c1"># Cell coordinates with metadata (pandas DataFrame)</span>
<span class="n">cells_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
    <span class="s1">&#39;x&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">x1</span><span class="p">,</span> <span class="n">x2</span><span class="p">,</span> <span class="o">...</span><span class="p">],</span>
    <span class="s1">&#39;y&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">y1</span><span class="p">,</span> <span class="n">y2</span><span class="p">,</span> <span class="o">...</span><span class="p">],</span>
    <span class="s1">&#39;cell_type&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;T-cell&#39;</span><span class="p">,</span> <span class="s1">&#39;B-cell&#39;</span><span class="p">,</span> <span class="o">...</span><span class="p">],</span>
    <span class="s1">&#39;gene_expression&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.7</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span>
<span class="p">})</span>

<span class="c1"># Image data (numpy array)</span>
<span class="n">img</span> <span class="o">=</span> <span class="n">tifffile</span><span class="o">.</span><span class="n">imread</span><span class="p">(</span><span class="s1">&#39;section_image.tif&#39;</span><span class="p">)</span>  <span class="c1"># Shape: (height, width) or (height, width, channels)</span>
</code></pre></div>
<h3 id="data-preprocessing">Data Preprocessing<a class="headerlink" href="#data-preprocessing" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Filter cells by type</span>
<span class="n">t_cells</span> <span class="o">=</span> <span class="n">cells_df</span><span class="p">[</span><span class="n">cells_df</span><span class="p">[</span><span class="s1">&#39;cell_type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;T-cell&#39;</span><span class="p">]</span>
<span class="n">t_cell_coords</span> <span class="o">=</span> <span class="n">t_cells</span><span class="p">[[</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="s1">&#39;y&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span>

<span class="c1"># Normalize coordinates</span>
<span class="n">normalized_points</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">compute</span><span class="o">.</span><span class="n">normalize_points</span><span class="p">(</span><span class="n">points</span><span class="p">)</span>

<span class="c1"># Process image data</span>
<span class="n">processed_img</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">img</span><span class="o">.</span><span class="n">preprocess_image</span><span class="p">(</span><span class="n">img</span><span class="p">,</span> <span class="n">contrast_enhance</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">denoise</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>
<h2 id="registration-workflow">Registration Workflow<a class="headerlink" href="#registration-workflow" title="Permanent link">&para;</a></h2>
<h3 id="full-registration-pipeline">Full Registration Pipeline<a class="headerlink" href="#full-registration-pipeline" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Load data for consecutive sections</span>
<span class="n">sections_data</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">):</span>  <span class="c1"># 5 sections</span>
    <span class="c1"># Load points and images</span>
    <span class="n">points</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;section</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">_points.npy&quot;</span><span class="p">)</span>
    <span class="n">img</span> <span class="o">=</span> <span class="n">tifffile</span><span class="o">.</span><span class="n">imread</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;section</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">_image.tif&quot;</span><span class="p">)</span>
    <span class="n">sections_data</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s2">&quot;points&quot;</span><span class="p">:</span> <span class="n">points</span><span class="p">,</span> <span class="s2">&quot;image&quot;</span><span class="p">:</span> <span class="n">img</span><span class="p">})</span>

<span class="c1"># Initialize transformation database</span>
<span class="n">transform_db</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">flow</span><span class="o">.</span><span class="n">outputs</span><span class="o">.</span><span class="n">TransformDB</span><span class="p">()</span>

<span class="c1"># Register consecutive sections</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sections_data</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
    <span class="n">source</span> <span class="o">=</span> <span class="n">sections_data</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
    <span class="n">target</span> <span class="o">=</span> <span class="n">sections_data</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span>

    <span class="c1"># Step 1: Affine registration</span>
    <span class="n">affine_transform</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">registration</span><span class="o">.</span><span class="n">affine_register</span><span class="p">(</span>
        <span class="n">source_points</span><span class="o">=</span><span class="n">source</span><span class="p">[</span><span class="s2">&quot;points&quot;</span><span class="p">],</span>
        <span class="n">target_points</span><span class="o">=</span><span class="n">target</span><span class="p">[</span><span class="s2">&quot;points&quot;</span><span class="p">]</span>
    <span class="p">)</span>

    <span class="c1"># Apply affine transformation</span>
    <span class="n">source_points_affine</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">affine</span><span class="o">.</span><span class="n">transform_points</span><span class="p">(</span><span class="n">source</span><span class="p">[</span><span class="s2">&quot;points&quot;</span><span class="p">],</span> <span class="n">affine_transform</span><span class="p">)</span>

    <span class="c1"># Step 2: Feature matching (optional, for improved alignment)</span>
    <span class="n">keypoints_source</span><span class="p">,</span> <span class="n">keypoints_target</span><span class="p">,</span> <span class="n">matches</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">matches</span><span class="o">.</span><span class="n">find_matches</span><span class="p">(</span>
        <span class="n">source</span><span class="p">[</span><span class="s2">&quot;image&quot;</span><span class="p">],</span> <span class="n">target</span><span class="p">[</span><span class="s2">&quot;image&quot;</span><span class="p">],</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;loftr&quot;</span>
    <span class="p">)</span>

    <span class="c1"># Step 3: Fine registration with flow</span>
    <span class="n">flow_transform</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">flow</span><span class="o">.</span><span class="n">register</span><span class="p">(</span>
        <span class="n">source_points</span><span class="o">=</span><span class="n">source_points_affine</span><span class="p">,</span>
        <span class="n">target_points</span><span class="o">=</span><span class="n">target</span><span class="p">[</span><span class="s2">&quot;points&quot;</span><span class="p">],</span>
        <span class="n">source_keypoints</span><span class="o">=</span><span class="n">keypoints_source</span><span class="p">,</span>
        <span class="n">target_keypoints</span><span class="o">=</span><span class="n">keypoints_target</span><span class="p">,</span>
        <span class="n">matches</span><span class="o">=</span><span class="n">matches</span><span class="p">,</span>
        <span class="n">iterations</span><span class="o">=</span><span class="mi">200</span>
    <span class="p">)</span>

    <span class="c1"># Store transforms in database</span>
    <span class="n">transform_db</span><span class="o">.</span><span class="n">add_transform</span><span class="p">(</span>
        <span class="n">source_id</span><span class="o">=</span><span class="n">i</span><span class="p">,</span>
        <span class="n">target_id</span><span class="o">=</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">affine_transform</span><span class="o">=</span><span class="n">affine_transform</span><span class="p">,</span>
        <span class="n">flow_transform</span><span class="o">=</span><span class="n">flow_transform</span>
    <span class="p">)</span>

<span class="c1"># Save transformation database</span>
<span class="n">transform_db</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;registration_results.pkl&quot;</span><span class="p">)</span>
</code></pre></div>
<h2 id="advanced-registration">Advanced Registration<a class="headerlink" href="#advanced-registration" title="Permanent link">&para;</a></h2>
<h3 id="multi-modal-registration">Multi-modal Registration<a class="headerlink" href="#multi-modal-registration" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Register using both point coordinates and image features</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">registration</span><span class="o">.</span><span class="n">multimodal_register</span><span class="p">(</span>
    <span class="n">source_points</span><span class="o">=</span><span class="n">source_points</span><span class="p">,</span>
    <span class="n">target_points</span><span class="o">=</span><span class="n">target_points</span><span class="p">,</span>
    <span class="n">source_image</span><span class="o">=</span><span class="n">source_image</span><span class="p">,</span>
    <span class="n">target_image</span><span class="o">=</span><span class="n">target_image</span><span class="p">,</span>
    <span class="n">point_weight</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span>  <span class="c1"># Weight for point-based alignment</span>
    <span class="n">image_weight</span><span class="o">=</span><span class="mf">0.3</span>   <span class="c1"># Weight for image-based alignment</span>
<span class="p">)</span>
</code></pre></div>
<h3 id="global-consistency-optimization">Global Consistency Optimization<a class="headerlink" href="#global-consistency-optimization" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Load transformation database</span>
<span class="n">transform_db</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">flow</span><span class="o">.</span><span class="n">outputs</span><span class="o">.</span><span class="n">TransformDB</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;registration_results.pkl&quot;</span><span class="p">)</span>

<span class="c1"># Optimize for global consistency</span>
<span class="n">optimized_transforms</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">registration</span><span class="o">.</span><span class="n">global_optimize</span><span class="p">(</span><span class="n">transform_db</span><span class="p">)</span>

<span class="c1"># Apply optimized transformations to all sections</span>
<span class="n">aligned_sections</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sections_data</span><span class="p">)):</span>
    <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># Reference section</span>
        <span class="n">aligned_sections</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sections_data</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;points&quot;</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Apply sequence of transformations to align to reference</span>
        <span class="n">aligned_points</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">registration</span><span class="o">.</span><span class="n">apply_transform_sequence</span><span class="p">(</span>
            <span class="n">sections_data</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s2">&quot;points&quot;</span><span class="p">],</span>
            <span class="n">optimized_transforms</span><span class="p">,</span>
            <span class="n">source_id</span><span class="o">=</span><span class="n">i</span><span class="p">,</span>
            <span class="n">target_id</span><span class="o">=</span><span class="mi">0</span>
        <span class="p">)</span>
        <span class="n">aligned_sections</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aligned_points</span><span class="p">)</span>
</code></pre></div>
<h2 id="working-with-large-datasets">Working with Large Datasets<a class="headerlink" href="#working-with-large-datasets" title="Permanent link">&para;</a></h2>
<h3 id="chunked-processing">Chunked Processing<a class="headerlink" href="#chunked-processing" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Process large point clouds in chunks</span>
<span class="n">chunk_size</span> <span class="o">=</span> <span class="mi">10000</span>  <span class="c1"># Number of points per chunk</span>
<span class="n">num_chunks</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">large_point_set</span><span class="p">)</span> <span class="o">//</span> <span class="n">chunk_size</span> <span class="o">+</span> <span class="mi">1</span>

<span class="n">transformed_points</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_chunks</span><span class="p">):</span>
    <span class="n">start_idx</span> <span class="o">=</span> <span class="n">i</span> <span class="o">*</span> <span class="n">chunk_size</span>
    <span class="n">end_idx</span> <span class="o">=</span> <span class="nb">min</span><span class="p">((</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">chunk_size</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">large_point_set</span><span class="p">))</span>

    <span class="n">chunk</span> <span class="o">=</span> <span class="n">large_point_set</span><span class="p">[</span><span class="n">start_idx</span><span class="p">:</span><span class="n">end_idx</span><span class="p">]</span>
    <span class="n">transformed_chunk</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">flow</span><span class="o">.</span><span class="n">transform_points</span><span class="p">(</span><span class="n">chunk</span><span class="p">,</span> <span class="n">flow_transform</span><span class="p">)</span>
    <span class="n">transformed_points</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">transformed_chunk</span><span class="p">)</span>

<span class="c1"># Combine all transformed chunks</span>
<span class="n">transformed_points</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">(</span><span class="n">transformed_points</span><span class="p">)</span>
</code></pre></div>
<h3 id="gpu-acceleration">GPU Acceleration<a class="headerlink" href="#gpu-acceleration" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Enable GPU acceleration</span>
<span class="n">sm</span><span class="o">.</span><span class="n">base</span><span class="o">.</span><span class="n">root</span><span class="o">.</span><span class="n">DEVICE</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># Use first GPU</span>

<span class="c1"># Perform registration with GPU acceleration</span>
<span class="n">flow_transform</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">flow</span><span class="o">.</span><span class="n">register</span><span class="p">(</span>
    <span class="n">source_points</span><span class="o">=</span><span class="n">source_points</span><span class="p">,</span>
    <span class="n">target_points</span><span class="o">=</span><span class="n">target_points</span><span class="p">,</span>
    <span class="n">use_gpu</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</code></pre></div>
<h2 id="visualizing-results">Visualizing Results<a class="headerlink" href="#visualizing-results" title="Permanent link">&para;</a></h2>
<h3 id="interactive-visualization">Interactive Visualization<a class="headerlink" href="#interactive-visualization" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Compare before/after registration</span>
<span class="n">sm</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">compare</span><span class="o">.</span><span class="n">interactive_comparison</span><span class="p">(</span>
    <span class="n">source_points</span><span class="o">=</span><span class="n">source_points</span><span class="p">,</span>
    <span class="n">target_points</span><span class="o">=</span><span class="n">target_points</span><span class="p">,</span>
    <span class="n">aligned_points</span><span class="o">=</span><span class="n">transformed_points</span>
<span class="p">)</span>

<span class="c1"># Visualize transformation field</span>
<span class="n">sm</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">show</span><span class="o">.</span><span class="n">plot_transformation_field</span><span class="p">(</span><span class="n">flow_transform</span><span class="p">)</span>

<span class="c1"># Interactive 3D model visualization</span>
<span class="n">sm</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">show</span><span class="o">.</span><span class="n">interactive_3d_plot</span><span class="p">(</span><span class="n">model_3d</span><span class="p">)</span>
</code></pre></div>
<h3 id="custom-visualization">Custom Visualization<a class="headerlink" href="#custom-visualization" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Create custom visualization with cell types</span>
<span class="n">cell_types</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;T-cell&#39;</span><span class="p">,</span> <span class="s1">&#39;B-cell&#39;</span><span class="p">,</span> <span class="s1">&#39;Epithelial&#39;</span><span class="p">,</span> <span class="s1">&#39;Stromal&#39;</span><span class="p">]</span>
<span class="n">cell_colors</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="s1">&#39;green&#39;</span><span class="p">,</span> <span class="s1">&#39;purple&#39;</span><span class="p">]</span>

<span class="c1"># Plot cells colored by type</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
<span class="k">for</span> <span class="n">cell_type</span><span class="p">,</span> <span class="n">color</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">cell_types</span><span class="p">,</span> <span class="n">cell_colors</span><span class="p">):</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="n">cells_df</span><span class="p">[</span><span class="s1">&#39;cell_type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">cell_type</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
        <span class="n">cells_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">mask</span><span class="p">,</span> <span class="s1">&#39;x&#39;</span><span class="p">],</span> 
        <span class="n">cells_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">mask</span><span class="p">,</span> <span class="s1">&#39;y&#39;</span><span class="p">],</span>
        <span class="n">c</span><span class="o">=</span><span class="n">color</span><span class="p">,</span>
        <span class="n">label</span><span class="o">=</span><span class="n">cell_type</span><span class="p">,</span>
        <span class="n">s</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
        <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span>
    <span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Cells by Type&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div>
<h2 id="exporting-3d-models">Exporting 3D Models<a class="headerlink" href="#exporting-3d-models" title="Permanent link">&para;</a></h2>
<h3 id="exporting-to-common-formats">Exporting to Common Formats<a class="headerlink" href="#exporting-to-common-formats" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Export 3D model to OBJ format</span>
<span class="n">sm</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">output</span><span class="o">.</span><span class="n">export_to_obj</span><span class="p">(</span><span class="n">model_3d</span><span class="p">,</span> <span class="s2">&quot;tissue_model.obj&quot;</span><span class="p">)</span>

<span class="c1"># Export as point cloud</span>
<span class="n">sm</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">output</span><span class="o">.</span><span class="n">export_point_cloud</span><span class="p">(</span><span class="n">model_3d</span><span class="p">,</span> <span class="s2">&quot;tissue_model.ply&quot;</span><span class="p">)</span>

<span class="c1"># Export to Imaris format</span>
<span class="n">sm</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">imaris</span><span class="o">.</span><span class="n">export_points</span><span class="p">(</span><span class="n">model_3d</span><span class="p">,</span> <span class="s2">&quot;tissue_model.ims&quot;</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</code></pre></div>
<h3 id="saving-results-for-analysis">Saving Results for Analysis<a class="headerlink" href="#saving-results-for-analysis" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Save aligned data for further analysis</span>
<span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;aligned_sections.npy&quot;</span><span class="p">,</span> <span class="n">aligned_sections</span><span class="p">)</span>

<span class="c1"># Export as CSV with metadata</span>
<span class="n">aligned_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
    <span class="s1">&#39;x&#39;</span><span class="p">:</span> <span class="n">model_3d</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span>
    <span class="s1">&#39;y&#39;</span><span class="p">:</span> <span class="n">model_3d</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span>
    <span class="s1">&#39;z&#39;</span><span class="p">:</span> <span class="n">model_3d</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">],</span>
    <span class="s1">&#39;section&#39;</span><span class="p">:</span> <span class="n">section_indices</span><span class="p">,</span>
    <span class="s1">&#39;cell_type&#39;</span><span class="p">:</span> <span class="n">cell_types</span>
<span class="p">})</span>
<span class="n">aligned_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;aligned_model.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div>
<h2 id="advanced-topics">Advanced Topics<a class="headerlink" href="#advanced-topics" title="Permanent link">&para;</a></h2>
<p>For more advanced usage, please refer to:</p>
<ul>
<li><a href="../../api/api/">API Reference</a> for detailed function documentation</li>
<li><a href="../examples/">Examples</a> for specific use cases</li>
<li>Individual module documentation for specialized features </li>
</ul>







  
    
  
  


  <aside class="md-source-file">
    
      
  <span class="md-source-file__fact">
    <span class="md-icon" title="Last update">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg>
    </span>
    <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date" title="May 10, 2025 13:04:30">May 10, 2025</span>
  </span>

    
    
    
    
  </aside>





                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.sections", "toc.integrate", "search.suggest", "search.highlight", "content.code.annotate"], "search": "../../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.13a4f30d.min.js"></script>
      
    
  </body>
</html>